src/cargo/ops/tree/mod.rs

```diff
@@ -0,0 +1,333 @@
+    if opts.no_filter_targets && opts.target.is_some() {
+        bail!("cannot specify both `--target` and `--no-filter-targets`");
+    }
+    if opts.graph_features && opts.duplicates {
+        bail!("the `--graph-features` flag does not support `--duplicates`");
+    }
+    let requested_kind = CompileKind::from_requested_target(ws.config(), opts.target.as_deref())?;
+    let target_data = RustcTargetData::new(ws, requested_kind)?;
+    let specs = opts.packages.to_package_id_specs(ws)?;
+    let resolve_opts = ResolveOpts::new(
+        /*dev_deps*/ true,
+        &opts.features,
+        opts.all_features,
+        !opts.no_default_features,
+    );
+    let has_dev = if opts.no_dev_dependencies {
+        HasDevUnits::No
+    } else {
+        HasDevUnits::Yes
+    };
+    let ws_resolve = ops::resolve_ws_with_opts(
+        ws,
+        &target_data,
+        requested_kind,
+        &resolve_opts,
+        &specs,
+        has_dev,
+    )?;
+    // Download all Packages. Some display formats need to display package metadata.
+    let package_map: HashMap<PackageId, &Package> = ws_resolve
+        .pkg_set
+        .get_many(ws_resolve.pkg_set.package_ids())?
+        .into_iter()
+        .map(|pkg| (pkg.package_id(), pkg))
+        .collect();
+
+    let mut graph = graph::build(
+        ws,
+        &ws_resolve.targeted_resolve,
+        &ws_resolve.resolved_features,
+        &specs,
+        &resolve_opts.features,
+        &target_data,
+        requested_kind,
+        package_map,
+        opts,
+    )?;
+
+    let root_ids = ws_resolve.targeted_resolve.specs_to_ids(&specs)?;
+    let root_indexes = graph.indexes_from_ids(&root_ids);
+
+    let root_indexes = if opts.duplicates {
+        // `-d -p foo` will only show duplicates within foo's subtree
+        graph = graph.from_reachable(root_indexes.as_slice());
+        graph.find_duplicates()
+    } else {
+        root_indexes
+    };
+
+    if opts.invert || opts.duplicates {
+        graph.invert();
+    }
+
+    print(opts, root_indexes, &graph)?;
+    Ok(())
+}
+
+/// Prints a tree for each given root.
+fn print(opts: &TreeOptions, roots: Vec<usize>, graph: &Graph<'_>) -> CargoResult<()> {
+    let format = Pattern::new(&opts.format)
+        .with_context(|| format!("tree format `{}` not valid", opts.format))?;
+
+    let symbols = match opts.charset {
+        Charset::Utf8 => &UTF8_SYMBOLS,
+        Charset::Ascii => &ASCII_SYMBOLS,
+    };
+
+    let prefix = if opts.prefix_depth {
+        Prefix::Depth
+    } else if opts.no_indent {
+        Prefix::None
+    } else {
+        Prefix::Indent
+    };
+
+    for (i, root_index) in roots.into_iter().enumerate() {
+        if i != 0 {
+            println!();
+        }
+
+        // The visited deps is used to display a (*) whenever a dep has
+        // already been printed (ignored with --no-dedupe).
+        let mut visited_deps = HashSet::new();
```

Comment:
Perhaps a minor point, but how come this is reset between each root?

Reply:
No good reason.  I changed it so that it de-dupes across roots.  I think that should be OK.